gordon/metrics/log.py

```diff
@@ -0,0 +1,127 @@
+# limitations under the License.
+
+import collections
+import logging
+import time
+
+import zope.interface
+
+from gordon import interfaces
+
+# TODO: update plugin loader to load metrics logger as default
+
+
+@zope.interface.implementer(interfaces.ITimer)
+class LogTimer:
+    """Timer that outputs metrics to the logger.
+
+    Args:
+        metric (dict): Metric to output.
+        logger (.MetricLogger): Object used for outputting metrics.
+        time_unit (number): (optional) Multiplier used to scale time.time()
+            output. For example, a value of `1` outputs whole seconds,
+            a value of `1000` outputs milliseconds.
+    """
+    def __init__(self, metric, logger, time_unit):
+        self.metric = metric
+        self.logger = logger
+        self.time_unit = time_unit
+
+    async def __aenter__(self):
+        await self.start()
+        return self
+
+    async def __aexit__(self, *args):
+        await self.stop()
+
+    async def start(self):
+        """Start the timer."""
+        self._start_time = time.time()
+
+    async def stop(self):
+        """Stop the timer and output the time delta to the logger."""
+        time_elapsed = (time.time() - self._start_time) * self.time_unit
+        self.metric['value'] = time_elapsed
+        self.logger.log(self.metric)
+
+
+class MetricLogger:
+    """Configurable wrapper that formats and outputs metrics.
+
+    Args:
+        level (str): Log level at which to output metric.
+        precision (str): Used to round metric to desired number of
+            places.
+    """
+    LOGFMT = '[{metric_name}] value: {value:.<prec>f}'
+
+    def __init__(self, level, precision):
+        self._log_fmt = self.LOGFMT.replace('<prec>', str(precision))
+        self._log_fn = getattr(logging, level.lower())
+
+    def log(self, metric):
+        """Format and output metric.
+
+        Args:
+            metric (dict): Complete metric.
+        """
+        message = self._log_fmt.format(**metric)
+        if metric['attrs']:
+            message += ' attrs: {attrs}'.format(attrs=metric['attrs'])
+        self._log_fn(message)
+
+
+@zope.interface.implementer(interfaces.IMetricRelay)
+class LogRelay:
+    """Manage metrics that get output to the logger.
+
+    Args:
+        config (dict): Required keys:
+            level (str): Log level of metrics log messages.
+            precision (int): Number of decimal places to round metric
+                value to for output.
+            time_unit (number): Scale unit to use with timers.
+    """
+    def __init__(self, config):
+        level = config.get('log_level', 'info')
+        precision = config.get('precision', 6)
+        self.logger = MetricLogger(level, precision)
+        self.time_unit = config.get('time_unit', 1)
+        self.counters = collections.defaultdict(int)
+
+    def _create_metric(self, metric_name, value, **kwargs):
+        attrs = kwargs.get('attrs') or {}
```

Comment:
I think we agreed to updated this stuff to `context` (perhaps that was after you made these changes, my apologies).